File: leetcode/problems/cpp/implement-trie-prefix-tree.cpp
// by @codeAbinash
// Time : O(n), n = length of word
// Space : O(n), 

#include <string>
using namespace std;

struct TrieNode {
   TrieNode* children[26]{};
   bool isEnd;
};

class Trie {
 public:
   TrieNode* root;

   Trie() { root = new TrieNode(); }

   void insert(string word) {
      TrieNode* node = root;
      for (char c : word) {
         if (node->children[c - 'a'] == nullptr) {
            node->children[c - 'a'] = new TrieNode();
         }
         node = node->children[c - 'a'];
      }
      node->isEnd = true;
   }

   bool search(string word) {
      TrieNode* node = root;
      for (char c : word) {
         if (node->children[c - 'a'] == nullptr) {
            return false;
         }
         node = node->children[c - 'a'];
      }
      return node->isEnd;
   }

   bool startsWith(string prefix) {
      TrieNode* node = root;
      for (char c : prefix) {
         if (node->children[c - 'a'] == nullptr) {
            return false;
         }
         node = node->children[c - 'a'];
      }
      return true;
   }
};

File: leetcode/problems/java/implement-trie-prefix-tree.java
class TrieNode {
  public TrieNode[] children = new TrieNode[26];
  public boolean isWord;
}

class Trie {

  TrieNode root;

  public Trie() {
    root = new TrieNode();
  }

  public void insert(String word) {
    TrieNode node = root;
    for (char c : word.toCharArray()) {
      if (node.children[c - 'a'] == null) {
        node.children[c - 'a'] = new TrieNode();
      }
      node = node.children[c - 'a'];
    }
    node.isWord = true;
  }

  public boolean search(String word) {
    TrieNode node = root;
    for (char c : word.toCharArray()) {
      if (node.children[c - 'a'] == null) {
        return false;
      }
      node = node.children[c - 'a'];
    }
    return node.isWord;
  }

  public boolean startsWith(String prefix) {
    TrieNode node = root;
    for (char c : prefix.toCharArray()) {
      if (node.children[c - 'a'] == null) {
        return false;
      }
      node = node.children[c - 'a'];
    }
    return true;
  }
}